From 4a264c7d82641e174f48aefd2fb83fcca2b5f8f5 Mon Sep 17 00:00:00 2001
From: J. Bruce Fields <bfields@redhat.com>
Date: Tue, 16 Sep 2014 20:19:46 -0400
Subject: [fs] nfsd: nfsd4_open_confirm() must reference the open stateid

Message-id: <1410898849-22126-103-git-send-email-bfields@redhat.com>
Patchwork-id: 95771
O-Subject: [Patch RHEL7 102/165] nfsd: nfsd4_open_confirm() must reference the open stateid
Bugzilla: 1044798 1123721 1141932
RH-Acked-by: Benjamin Coddington <bcodding@redhat.com>
RH-Acked-by: Steve Dickson <SteveD@redhat.com>

From: Trond Myklebust <trond.myklebust@primarydata.com>

Ensure that nfsd4_open_confirm() keeps a reference to the open
stateid until it is done working with it.

Necessary step toward client_mutex removal.

Signed-off-by: Trond Myklebust <trond.myklebust@primarydata.com>
Signed-off-by: J. Bruce Fields <bfields@redhat.com>
Upstream: 2585fc79584684666b3d107179e43484dfb4da13
Bugzilla: 1044798
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/fs/nfsd/nfs4state.c b/fs/nfsd/nfs4state.c
index 4e50f14..8e18ca4 100644
--- a/fs/nfsd/nfs4state.c
+++ b/fs/nfsd/nfs4state.c
@@ -4434,10 +4434,12 @@ nfsd4_open_confirm(struct svc_rqst *rqstp, struct nfsd4_compound_state *cstate,
      NFS4_OPEN_STID, &stp, nn);
  if (status)
   goto out;
+ /* FIXME: move into nfs4_preprocess_seqid_op */
+ atomic_inc(&stp->st_stid.sc_count);
  oo = openowner(stp->st_stateowner);
  status = nfserr_bad_stateid;
  if (oo->oo_flags & NFS4_OO_CONFIRMED)
-  goto out;
+  goto put_stateid;
  oo->oo_flags |= NFS4_OO_CONFIRMED;
  update_stateid(&stp->st_stid.sc_stateid);
  memcpy(&oc->oc_resp_stateid, &stp->st_stid.sc_stateid, sizeof(stateid_t));
@@ -4446,6 +4448,8 @@ nfsd4_open_confirm(struct svc_rqst *rqstp, struct nfsd4_compound_state *cstate,
 
  nfsd4_client_record_create(oo->oo_owner.so_client);
  status = nfs_ok;
+put_stateid:
+ nfs4_put_stid(&stp->st_stid);
 out:
  nfsd4_bump_seqid(cstate, status);
  if (!cstate->replay_owner)
-- 
1.7.1